Added takewhile

reinermartin opened the pull request:
Added a takewhile function as discussed in issue "Would takewhile be a natural candidate for IterTools.jl?"

ararslan commented on src/IterTools.jl:
```diff
@@ -873,4 +873,35 @@ function iterate(ff::FlagFirst, state = (true, ))
+ 2
+ 3
+"""
+
```
There can't be extra blank lines between a method and its docstring anymore as of Julia 0.7, so this line should be removed.

reinermartin replied:
OK, should be fixed now

ararslan commented:
The test failures should be fixed by simply adding `takewhile` to the list of `export`s.

reinermartin commented:
> The test failures should be fixed by simply adding `takewhile` to the list of `export`s.



> The test failures should be fixed by simply adding `takewhile` to the list of `export`s.

Thanks, done that as well now

ararslan commented:
Great work here!

reinermartin commented:
> Great work here!

Thanks for your help!

iamed2 commented on src/IterTools.jl:
```diff
@@ -873,4 +874,34 @@ function iterate(ff::FlagFirst, state = (true, ))
+takewhile(cond, xs) = TakeWhile(cond, xs)
+
+function Base.iterate(it::TakeWhile, state=nothing)
+    (val, state) = state == nothing ? iterate(it.xs) : iterate(it.xs, state)
```
Should be `state === nothing`

iamed2 commented on src/IterTools.jl:
```diff
@@ -873,4 +874,34 @@ function iterate(ff::FlagFirst, state = (true, ))
+    val, state
+end
+
+Base.IteratorSize(it::TakeWhile) = Base.SizeUnknown()
```
Define `IteratorEltype` and `eltype` as well, like I did in `Ivec`. Then you won't get `Any` arrays returned.

reinermartin commented:
iamed2, all your suggestion should be implemented, plus I fixed a bug (when condition was always true for a finite iterator it failed, I added @someting to fix this)
